tests are failing on the clean install

Reproducer:
```bash
mkdir -p $(go env GOPATH)/src/github.com/go-critic
cd $(go env GOPATH)/src/github.com/go-critic
git clone https://github.com/go-critic/go-critic.git
dep ensure
go test -v -count=1 ./lint
```
Tests will fail due to lack of `astfmt` package:
```
$GOPATH/src/github.com/go-critic/go-critic/lint/testdata/importShadow/negative_tests.go:7:4: could not import github.com/go-toolsmith/astfmt (cannot find package "github.com/go-toolsmith/astfmt" in any of:
	$GOROOT/src/github.com/go-toolsmith/astfmt (from $GOROOT)
	$GOPATH/src/github.com/go-toolsmith/astfmt (from $GOPATH))
```

--- a/Gopkg.lock
+++ b/Gopkg.lock
@@ -3,47 +3,82 @@
 
 [[projects]]
   branch = "master"
+  digest = "1:71ce2e3f3b4acb600c5bbcffccd0d11534cec4e29a73f6de5781bae11ec528a3"
   name = "github.com/go-toolsmith/astcopy"
   packages = ["."]
-  revision = "fc497f942afe9ca951d5c91c5c180fd59aebe5d4"
+  pruneopts = "UT"
+  revision = "79b422d080c469383ec3cf020d124b70d073956b"
 
 [[projects]]
   branch = "master"
+  digest = "1:db8123e0896cf805cb30c7bacad0b1a1d562edbbcd2a765ceda1af926c9a99a4"
   name = "github.com/go-toolsmith/astequal"
   packages = ["."]
-  revision = "e9d31615533153b6a0f02f491c5432b227094163"
+  pruneopts = "UT"
+  revision = "dcb477bfacd6e00a13c6d63bfc73db28dd343160"
 
 [[projects]]
   branch = "master"
+  digest = "1:f4fcc4dafc60c8612e7b8855310f0f7f5b7b6947011e60f26fb9c10a690b0b6f"
   name = "github.com/go-toolsmith/astfmt"
   packages = ["."]
-  revision = "6ea5b99ab1746954c14cc74d79d94913a53c8ff3"
+  pruneopts = "UT"
+  revision = "8f8ee99c30864ae6d51ec6c8454d72a5072fe703"
 
 [[projects]]
   branch = "master"
+  digest = "1:04989378ee2deb38e1f82fb5ec962a789f675e394c42c9df79fcc8f4e47d9005"
   name = "github.com/go-toolsmith/astp"
   packages = ["."]
-  revision = "d378e1e6ba8e7084e724d9aa4ce0b41b719d5257"
+  pruneopts = "UT"
+  revision = "0af7e3c24f3087820a2e96679d064e6a56c624e2"
 
 [[projects]]
   branch = "master"
+  digest = "1:aba9fa7c6dacad505e95d984ec0b07f6b5a4691ed2f4ff36c7b732b3704627a3"
   name = "github.com/go-toolsmith/strparse"
   packages = ["."]
-  revision = "71162a977d98e0f84f53e3b1424e54a9a3b615ef"
+  pruneopts = "UT"
+  revision = "830b6daa1241714c12a9b9a4a56849fe2f93aedc"
 
 [[projects]]
   branch = "master"
+  digest = "1:be40f095cd741773905744f16c1f7a21fd9226ccd0529f019deb7da6d667f71c"
+  name = "github.com/logrusorgru/aurora"
+  packages = ["."]
+  pruneopts = "UT"
+  revision = "a7b3b318ed4e1ae5b80602b08627267303c68572"
+
+[[projects]]
+  branch = "master"
+  digest = "1:27e7fb3f0e273bb9b88b6f76d056f76f6a57401f43bd2bf2eabd24a61cefe558"
   name = "golang.org/x/tools"
   packages = [
     "go/ast/astutil",
-    "go/buildutil",
-    "go/loader"
+    "go/gcexportdata",
+    "go/internal/cgo",
+    "go/internal/gcimporter",
+    "go/packages",
+    "go/types/typeutil",
+    "internal/fastwalk",
+    "internal/gopathwalk",
+    "internal/semver",
   ]
-  revision = "56e9b8e653c818f46a96eb4828aba7be54271a69"
+  pruneopts = "UT"
+  revision = "f60e5f99f0816fc2d9ecb338008ea420248d2943"
 
 [solve-meta]
   analyzer-name = "dep"
   analyzer-version = 1
-  inputs-digest = "09adb9af9c3491ef5fc14d91628f9f7c0ad6894f09bf93ea24db0c5e287e61f9"
+  input-imports = [
+    "github.com/go-toolsmith/astcopy",
+    "github.com/go-toolsmith/astequal",
+    "github.com/go-toolsmith/astfmt",
+    "github.com/go-toolsmith/astp",
+    "github.com/go-toolsmith/strparse",
+    "github.com/logrusorgru/aurora",
+    "golang.org/x/tools/go/ast/astutil",
+    "golang.org/x/tools/go/packages",
+  ]
   solver-name = "gps-cdcl"
   solver-version = 1

--- a/Gopkg.toml
+++ b/Gopkg.toml
@@ -1,6 +1,6 @@
 # Gopkg.toml example
 #
-# Refer to https://github.com/golang/dep/blob/master/docs/Gopkg.toml.md
+# Refer to https://golang.github.io/dep/docs/Gopkg.toml.html
 # for detailed Gopkg.toml documentation.
 #
 # required = ["github.com/user/thing/cmd/thing"]
@@ -44,6 +44,10 @@
 [[constraint]]
   branch = "master"
   name = "github.com/go-toolsmith/strparse"
+
+[[constraint]]
+  branch = "master"
+  name = "github.com/logrusorgru/aurora"
 
 [[constraint]]
   branch = "master"

--- a/cmd/criticize/main.go
+++ b/cmd/criticize/main.go
@@ -164,31 +164,58 @@ func (l *linter) LoadProgram() {
 	}
 
 	l.fset = token.NewFileSet()
-	conf := packages.Config{
+	cfg := packages.Config{
 		Mode:  packages.LoadSyntax,
 		Tests: true,
 		Fset:  l.fset,
 	}
-
-	pkgs, err := packages.Load(&conf, l.packages...)
+	pkgs, err := loadPackages(&cfg, l.packages)
 	if err != nil {
 		log.Fatalf("load packages: %v", err)
 	}
+	sort.SliceStable(pkgs, func(i, j int) bool {
+		return pkgs[i].PkgPath < pkgs[j].PkgPath
+	})
 
+	l.loadedPackages = pkgs
+	l.ctx = lint.NewContext(l.fset, sizes)
+}
+
+func loadPackages(cfg *packages.Config, patterns []string) ([]*packages.Package, error) {
+	pkgs, err := packages.Load(cfg, patterns...)
+	if err != nil {
+		return nil, err
+	}
+
+	result := pkgs[:0]
+
+	// First pkgs traversal selects external tests and
+	// packages built for testing.
+	// If there is no tests for the package,
+	// we're going to check them during the second traversal
+	// which visits normal package if only it was
+	// not checked during the first traversal.
+	withTests := map[string]bool{}
 	for _, pkg := range pkgs {
-		// For some patterns Load returns 4 packages.
-		// We need at most 2 and both of them should
-		// have [$pkg.test] parts in their ID.
 		if !strings.Contains(pkg.ID, ".test]") {
 			continue
 		}
-		l.loadedPackages = append(l.loadedPackages, pkg)
+		result = append(result, pkg)
+		withTests[pkg.PkgPath] = true
 	}
-	sort.SliceStable(l.loadedPackages, func(i, j int) bool {
-		return l.loadedPackages[i].PkgPath < l.loadedPackages[j].PkgPath
-	})
+	for _, pkg := range pkgs {
+		if strings.HasSuffix(pkg.PkgPath, ".test") {
+			continue
+		}
+		if pkg.ID != pkg.PkgPath {
+			continue
+		}
+		if !withTests[pkg.PkgPath] {
+			result = append(result, pkg)
+		}
+	}
 
-	l.ctx = lint.NewContext(conf.Fset, sizes)
+	return result, nil
 }
 
 func (l *linter) InitCheckers() {

--- a/cmd/criticize/main.go
+++ b/cmd/criticize/main.go
@@ -5,7 +5,7 @@ import (
 	"flag"
 	"go/ast"
 	"go/build"
-	"go/parser"
+	"go/token"
 	"go/types"
 	"io/ioutil"
 	"log"
@@ -13,26 +13,34 @@ import (
 	"path/filepath"
 	"regexp"
 	"runtime"
+	"sort"
 	"strings"
 	"sync"
 
 	"github.com/go-critic/go-critic/cmd/internal/flagparser"
 	"github.com/go-critic/go-critic/lint"
 	"github.com/logrusorgru/aurora"
-	"golang.org/x/tools/go/loader"
+	"golang.org/x/tools/go/packages"
 )
 
 var generatedFileCommentRE = regexp.MustCompile("Code generated .* DO NOT EDIT.")
 
+type resolvedPath struct {
+	path         string
+	pkg          *packages.Package
+	externalTest *packages.Package
+}
+
 type linter struct {
 	ctx *lint.Context
 
-	prog *loader.Program
+	loadedPackages []*packages.Package
 
 	checkers []*lint.Checker
 
 	foundIssues bool // True if there any checker reported an issue
 
+	fset  *token.FileSet
 	flags *flagparser.FlagParser
 
 	packages        []string
@@ -97,22 +105,8 @@ func Main() {
 	l.LoadProgram()
 	l.InitCheckers()
 
-	pkgInfoMap := make(map[string]*loader.PackageInfo)
-	for _, pkgInfo := range l.prog.AllPackages {
-		pkgInfoMap[pkgInfo.Pkg.Path()] = pkgInfo
-	}
-	for _, pkgPath := range l.packages {
-		pkgInfo := pkgInfoMap[pkgPath]
-		if pkgInfo == nil || !pkgInfo.TransitivelyErrorFree {
-			log.Fatalf("%s package is not properly loaded", pkgPath)
-		}
-		// Check the package itself.
-		l.CheckPackage(pkgPath, pkgInfo)
-		// Check package external test (if any).
-		pkgInfo = pkgInfoMap[pkgPath+"_test"]
-		if pkgInfo != nil {
-			l.CheckPackage(pkgPath+"_test", pkgInfo)
-		}
+	for _, pkg := range l.loadedPackages {
+		l.CheckPackage(pkg)
 	}
 
 	os.Exit(l.ExitCode())
@@ -175,23 +169,32 @@ func (l *linter) LoadProgram() {
 		log.Fatalf("can't find sizes info for %s", runtime.GOARCH)
 	}
 
-	conf := loader.Config{
-		ParserMode: parser.ParseComments,
-		TypeChecker: types.Config{
-			Sizes: sizes,
-		},
+	l.fset = token.NewFileSet()
+	conf := packages.Config{
+		Mode:  packages.LoadSyntax,
+		Tests: true,
+		Fset:  l.fset,
 	}
 
-	if _, err := conf.FromArgs(l.packages, true); err != nil {
-		log.Fatalf("resolve packages: %v", err)
-	}
-	prog, err := conf.Load()
+	pkgs, err := packages.Load(&conf, l.packages...)
 	if err != nil {
-		log.Fatalf("load program: %v", err)
+		log.Fatalf("load packages: %v", err)
 	}
 
-	l.prog = prog
-	l.ctx = lint.NewContext(prog.Fset, sizes)
+	for _, pkg := range pkgs {
+		// For some patterns Load returns 4 packages.
+		// We need at most 2 and both of them should
+		// have [$pkg.test] parts in their ID.
+		if !strings.Contains(pkg.ID, ".test]") {
+			continue
+		}
+		l.loadedPackages = append(l.loadedPackages, pkg)
+	}
+	sort.SliceStable(l.loadedPackages, func(i, j int) bool {
+		return l.loadedPackages[i].PkgPath < l.loadedPackages[j].PkgPath
+	})
+
+	l.ctx = lint.NewContext(conf.Fset, sizes)
 }
 
 func (l *linter) InitCheckers() {
@@ -235,9 +238,9 @@ func (l *linter) InitCheckers() {
 	}
 }
 
-func (l *linter) CheckPackage(pkgPath string, pkgInfo *loader.PackageInfo) {
-	l.ctx.SetPackageInfo(&pkgInfo.Info, pkgInfo.Pkg)
-	for _, f := range pkgInfo.Files {
+func (l *linter) CheckPackage(pkg *packages.Package) {
+	l.ctx.SetPackageInfo(pkg.TypesInfo, pkg.Types)
+	for _, f := range pkg.Syntax {
 		if l.flags.IgnoreTests && l.isTestFile(f) {
 			continue
 		}
@@ -258,8 +261,7 @@ func (l *linter) isGenerated(f *ast.File) bool {
 }
 
 func (l *linter) getFilename(f *ast.File) string {
-	// see https://github.com/golang/go/issues/24498
-	return filepath.Base(l.prog.Fset.Position(f.Pos()).Filename)
+	return filepath.Base(l.fset.Position(f.Pos()).Filename)
 }
 
 // ExitCode returns status code that should be used as an argument to os.Exit.

--- a/cmd/criticize/main.go
+++ b/cmd/criticize/main.go
@@ -25,12 +25,6 @@ import (
 
 var generatedFileCommentRE = regexp.MustCompile("Code generated .* DO NOT EDIT.")
 
-type resolvedPath struct {
-	path         string
-	pkg          *packages.Package
-	externalTest *packages.Package
-}
-
 type linter struct {
 	ctx *lint.Context
 

--- a/lint/all_checkers_test.go
+++ b/lint/all_checkers_test.go
@@ -3,7 +3,6 @@ package lint
 import (
 	"fmt"
 	"go/ast"
-	"go/parser"
 	"go/types"
 	"io/ioutil"
 	"os"
@@ -14,7 +13,7 @@ import (
 	"strings"
 	"testing"
 
-	"golang.org/x/tools/go/loader"
+	"golang.org/x/tools/go/packages"
 )
 
 var (
@@ -50,13 +49,11 @@ func TestSanity(t *testing.T) {
 		t.Run(rule.Name(), func(t *testing.T) {
 			pkgPath := testdataPkgPath + "/_sanity"
 
-			prog := newProg(t, pkgPath)
-			pkgInfo := prog.Imported[pkgPath]
+			pkg := newPackage(t, pkgPath)
 
-			ctx := NewContext(prog.Fset, sizes)
-			ctx.SetPackageInfo(&pkgInfo.Info, pkgInfo.Pkg)
-
-			files := prog.Imported[pkgPath].Files
+			ctx := NewContext(pkg.Fset, sizes)
+			ctx.SetPackageInfo(pkg.TypesInfo, pkg.Types)
+			files := pkg.Syntax
 
 			for _, f := range files {
 				defer func() {
@@ -85,22 +82,20 @@ func TestCheckers(t *testing.T) {
 			}
 			pkgPath := testdataPkgPath + testRule.Name()
 
-			prog := newProg(t, pkgPath)
-			pkgInfo := prog.Imported[pkgPath]
+			pkg := newPackage(t, pkgPath)
+			ctx := NewContext(pkg.Fset, sizes)
+			ctx.SetPackageInfo(pkg.TypesInfo, pkg.Types)
 
-			ctx := NewContext(prog.Fset, sizes)
-			ctx.SetPackageInfo(&pkgInfo.Info, pkgInfo.Pkg)
-
-			checkFiles(t, testRule, ctx, prog, pkgPath)
+			checkFiles(t, testRule, ctx, pkg, pkgPath)
 		})
 	}
 }
 
-func checkFiles(t *testing.T, rule *Rule, ctx *Context, prog *loader.Program, pkgPath string) {
-	files := prog.Imported[pkgPath].Files
+func checkFiles(t *testing.T, rule *Rule, ctx *Context, pkg *packages.Package, pkgPath string) {
+	files := pkg.Syntax
 
 	for _, f := range files {
-		filename := getFilename(prog, f)
+		filename := getFilename(pkg, f)
 		testFilename := filepath.Join("testdata", rule.Name(), filename)
 		goldenWarns := newGoldenFile(t, testFilename)
 
@@ -174,9 +169,9 @@ func TestIncorrectRule(t *testing.T) {
 	}(t)
 }
 
-func getFilename(prog *loader.Program, f *ast.File) string {
+func getFilename(pkg *packages.Package, f *ast.File) string {
 	// see https://github.com/golang/go/issues/24498
-	return filepath.Base(prog.Fset.Position(f.Pos()).Filename)
+	return filepath.Base(pkg.Fset.Position(f.Pos()).Filename)
 }
 
 type goldenFile struct {
@@ -239,23 +234,16 @@ func (f *goldenFile) checkUnmatched(t *testing.T, testFilename string) {
 	}
 }
 
-func newProg(t *testing.T, pkgPath string) *loader.Program {
-	conf := loader.Config{
-		ParserMode: parser.ParseComments,
-		TypeChecker: types.Config{
-			Sizes: sizes,
-		},
-	}
-	if _, err := conf.FromArgs([]string{pkgPath}, true); err != nil {
-		t.Fatalf("resolve packages: %v", err)
-	}
-	prog, err := conf.Load()
+func newPackage(t *testing.T, pkgPath string) *packages.Package {
+	conf := &packages.Config{Mode: packages.LoadSyntax}
+	pkgs, err := packages.Load(conf, pkgPath)
+
 	if err != nil {
 		t.Fatal(err)
 	}
-	pkgInfo := prog.Imported[pkgPath]
-	if pkgInfo == nil || !pkgInfo.TransitivelyErrorFree {
-		t.Fatalf("%s package is not properly loaded", pkgPath)
+	if len(pkgs) != 1 {
+		t.Fatalf("more than 1 packages loaded for %s", pkgPath)
 	}
-	return prog
+
+	return pkgs[0]
 }

